source/server/http_dynamic_delay_filter.h

```diff
@@ -0,0 +1,98 @@
+#pragma once
+
+#include <atomic>
+#include <string>
+
+#include "envoy/server/filter_config.h"
+
+#include "api/server/response_options.pb.h"
+
+namespace Nighthawk {
+namespace Server {
+
+/**
+ * Filter configuration container class for the dynamic delay extension.
+ */
+class HttpDynamicDelayDecoderFilterConfig {
+
+public:
+  HttpDynamicDelayDecoderFilterConfig(nighthawk::server::ResponseOptions proto_config);
+
+  /**
+   * @return const nighthawk::server::ResponseOptions& read-only reference to the proto config
+   * object.
+   */
+  const nighthawk::server::ResponseOptions& server_config() { return server_config_; }
+
+  /**
+   * Increments the number of active instances.
+   */
+  void incrementInstanceCount() { instances()++; }
+  /**
+   * Decrements the number of active instances.
+   */
+  void decrementInstanceCount() { instances()--; }
+  /**
+   * @return uint64_t the approximate number of active instances.
+   */
+  uint64_t approximateInstances() const { return instances(); }
+
+private:
+  const nighthawk::server::ResponseOptions server_config_;
+  static std::atomic<uint64_t>& instances() {
+    // We lazy-init the atomic to avoid static initialization / a fiasco.
+    MUTABLE_CONSTRUCT_ON_FIRST_USE(std::atomic<uint64_t>, 0); // NOLINT
+  }
+};
+
+using HttpDynamicDelayDecoderFilterConfigSharedPtr =
+    std::shared_ptr<HttpDynamicDelayDecoderFilterConfig>;
+
+/**
+ * Extension that controls the fault filter extension by supplying it with a request
+ * header that triggers it to induce a delay under the hood.
+ * In the future, we may look into injecting the fault filter ourselves with the right
+ * configuration, either directly/verbatim, or by including a derivation of it in
+ * this code base, thereby making it all transparant to the user / eliminating the need
+ * to configure the fault filter and make NH take full ownership at the feature level.
+ */
+class HttpDynamicDelayDecoderFilter : public Envoy::Http::StreamDecoderFilter {
+public:
+  HttpDynamicDelayDecoderFilter(HttpDynamicDelayDecoderFilterConfigSharedPtr);
+  ~HttpDynamicDelayDecoderFilter() override;
+
+  // Http::StreamFilterBase
+  void onDestroy() override;
+
+  // Http::StreamDecoderFilter
+  Envoy::Http::FilterHeadersStatus decodeHeaders(Envoy::Http::RequestHeaderMap&, bool) override;
```

Comment:
These appear to be overriding, and then just passing a CONTINUE, almost as if they're a no-op.  Could you add a comment explaining why they are doing that?

Reply:
We're implementing a pure virtual interface [1], so there's no real choice here, we must override / implement.
Since we don't have anything useful to do in these implementations, we indicate we'd like to continue to the state machine driving this. Indeed a no-op, but I'm also not sure I can add a real helpful comment about it?

[1] https://github.com/envoyproxy/envoy/blob/master/include/envoy/http/filter.h#L515

Reply:
understood.  thanks for explaining.  i think what i'm looking for is a just a quick comment to help readers know that they don't have to spend their time understanding what these are, but if you don't think that's necessary, i'll follow your lead.

Reply:
Done;